### PR TITLE
[solvers] Reformulate SolverOptions.SetOption to use our variant

### PR DESCRIPTION
**Pull request by jwnimmer-tri**
This simplifies our API surface for users, and makes the implementation of #22078 easier.

Towards #22078.

- - -
This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22282)

### CODE REVIEW

**Comment by jwnimmer-tri** [comment]
+@hongkai-dai would you like to feature-review this?

**Review by hongkai-dai — COMMENTED** [review]
> Reviewed 4 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22282)* status: 1 unresolved discussion, LGTM missing from assignee hongkai-dai, needs platform reviewer assigned, needs at least two assigned reviewers

___
*[`solvers/solver_options.cc` line 28 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/22282#-ODh9rCWD6eDCnbS6a95:-ODh9rCWD6eDCnbS6a96:bu0go81) ([raw file](https://github.com/RobotLocomotion/drake/blob/1bede164b6e45ffbb56b25f35055a90783b63d1f/solvers/solver_options.cc#L28)):*
> ```cpp
>   std::visit(
>       overloaded{
>           [&](double& unboxed_value) {
> ```

Curious why we use a reference `double&` instead of const reference here? And why we use `std:move(key)` instead of `const std::string_view& key`?

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22282)* status: 1 unresolved discussion, LGTM missing from assignee hongkai-dai, needs platform reviewer assigned, needs at least two assigned reviewers

___
*[`solvers/solver_options.cc` line 28 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/22282#-ODh9rCWD6eDCnbS6a95:-ODhEnG9Bt31KrWiceEs:b-x3gdoq) ([raw file](https://github.com/RobotLocomotion/drake/blob/1bede164b6e45ffbb56b25f35055a90783b63d1f/solvers/solver_options.cc#L28)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Curious why we use a reference `double&` instead of const reference here? And why we use `std:move(key)` instead of `const std::string_view& key`?
</blockquote></details>

The goal is to avoid move the `value` into place when it's a `string`.  I've respelled the code a bit to hopefully make that more clear?

**Review by hongkai-dai — COMMENTED** [review]
> Reviewed 1 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22282)* status: 1 unresolved discussion, LGTM missing from assignee hongkai-dai, needs platform reviewer assigned, needs at least two assigned reviewers

___
*[`solvers/solver_options.cc` line 28 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/22282#-ODh9rCWD6eDCnbS6a95:-ODj4ojD7Q-91Mo1mDGX:b-aiwalc) ([raw file](https://github.com/RobotLocomotion/drake/blob/1bede164b6e45ffbb56b25f35055a90783b63d1f/solvers/solver_options.cc#L28)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The goal is to avoid move the `value` into place when it's a `string`.  I've respelled the code a bit to hopefully make that more clear?
</blockquote></details>

I see. Thanks! Let me rephrase it to make sure I understand

Since we just move `value` into `solver_options_str_`, we want to make sure that everything is rvalue, hence we use rvalue reference here.

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22282)* status: 1 unresolved discussion, LGTM missing from assignee hongkai-dai, needs platform reviewer assigned, needs at least two assigned reviewers

___
*[`solvers/solver_options.cc` line 28 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/22282#-ODh9rCWD6eDCnbS6a95:-ODlFiUj39HqpquLTxJX:bmrv20b) ([raw file](https://github.com/RobotLocomotion/drake/blob/1bede164b6e45ffbb56b25f35055a90783b63d1f/solvers/solver_options.cc#L28)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I see. Thanks! Let me rephrase it to make sure I understand

Since we just move `value` into `solver_options_str_`, we want to make sure that everything is rvalue, hence we use rvalue reference here.
</blockquote></details>

Yes, exactly.

(I believe using the lvalue reference and moving from it (as I did in r1) would also still avoid the copying, but is less clear to a person reading the code.  Writing out the `&&` in the signature here most clearly communicate the intention.)

**Review by hongkai-dai — COMMENTED** [review]
:lgtm: +@ggould-tri for platform review please, thanks!

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22282#-:-ODlQ5c9C2iV4F8D1pKh:b-svovrs)* status: LGTM missing from assignee ggould-tri(platform)

___
*[`solvers/solver_options.cc` line 28 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/22282#-ODh9rCWD6eDCnbS6a95:-ODlPzzj86sZRLx9cNtL:bx40oul) ([raw file](https://github.com/RobotLocomotion/drake/blob/1bede164b6e45ffbb56b25f35055a90783b63d1f/solvers/solver_options.cc#L28)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Yes, exactly.

(I believe using the lvalue reference and moving from it (as I did in r1) would also still avoid the copying, but is less clear to a person reading the code.  Writing out the `&&` in the signature here most clearly communicate the intention.)
</blockquote></details>

Got it, thanks for the explanation!

**Review by ggould-tri — COMMENTED** [review]
:lgtm:

> Reviewed 4 of 5 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22282#-:-ODlUIYr25Far1CV0mp7:bnfp4nl)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignees ggould-tri(platform),hongkai-dai
